wagtail/admin/rich_text/converters/contentstate.py

```diff
@@ -0,0 +1,94 @@
+import json
+import logging
+import re
+
+from draftjs_exporter.defaults import render_children
+from draftjs_exporter.dom import DOM
+from draftjs_exporter.html import HTML as HTMLExporter
+
+from wagtail.admin.rich_text.converters.contentstate_models import Block
+from wagtail.admin.rich_text.converters.html_to_contentstate import HtmlToContentStateHandler
+from wagtail.core.rich_text import features as feature_registry
+
+
+def link_entity(props):
+    """
+    <a linktype="page" id="1">internal page link</a>
+    """
+    id_ = props.get('id')
+    link_props = {}
+
+    if id_ is not None:
+        link_props['linktype'] = 'page'
+        link_props['id'] = id_
+    else:
+        link_props['href'] = props.get('url')
+
+    return DOM.create_element('a', link_props, props['children'])
+
+
+def br(props):
+    if props['block']['type'] == 'code-block':
```

Comment:
For composite decorators, the `block` prop refers to the block the decorator is called inside of. The strategy matches all `\n` because it's based on text only, so the decorator has to skip the transformation if the block happens to be code.